### DISCUSSION TITLE
How to `dev` with `prerender-data-provider`?

### DISCUSSION THREAD

**Discussion by orangecoloured** [preactjs/preact-cli#1802] [Q&A] [answered]
I decided to use `prerender-data-provider` to load a JSON for a specific route.
I added the `<Provider value={props}>` wrapper in the `App` component.

But when I run the app I get an error `Error: The prerender-data provider expects current URL in props. This is required to give the pre-render data to the correct instance only.`
And when I look at the `props` contents in the `App` component, it's `{CLI_DATA: {preRenderData: {}}}`.

What did I miss in the setup?

**Comment by rschristian** [comment]
Did you set `--prerender`?

Prerendering in dev will only run on first build, you will not get refreshed prerender data/HTML without restarting the dev server, but this will provide the prerender data that you can use in your app.

**Reply by orangecoloured** [reply]
Do you mean `yarn dev` should run `preact watch --prerender`?

**Reply by rschristian** [reply]
Yes

https://github.com/preactjs/preact-cli#preact-watch

**Reply by orangecoloured** [reply]
Thanks. I was reading the docs [here](https://preactjs.com/guide/v10/cli/pre-rendering), the flag is not mentioned there.

Now `props` is `{CLI_DATA: {preRenderData: {url: "/"}}}` on the `App` component level. And it's always the same no matter the actual route. Also `props` inside any of the `src/routes` components is empty, `{}`.

**Reply by rschristian** [reply]
I wouldn't reference those pages for CLI information, it's simply not kept up-to-date.

> Now props is {CLI_DATA: {preRenderData: {url: "/"}}} on the App component level. And it's always the same no matter the actual route. Also props inside any of the src/routes components is empty, {}.

I'm assuming you have set other routes to be prerendered in your `prerender-urls.json`? If so, keep in mind that the prerender data comes from the HTML document that you load, not routes. They're completely separate things.

If I open `localhost:8080/`, I'll get an HTML doc with the URL set to `/`; navigating to a new URL with client-side routing will not change this. You'll have to navigate to a separate HTML doc (say `/foo.html`) using browser navigation instead (so as to not use the client-side router).

**Reply by orangecoloured** [reply]
Yes, my data comes from `prerender-urls.js`.

But if I load `localhost:8080/something`, should I get the `{ url: '/something' }`?

Sorry, I'm a bit confused with how the prerender provider works. Can I use the data from `prerender-urls.js` in dev? And if I can't, how can I substitute it? What is the way to mimic the prerender behaviour while in dev mode?

**Reply by rschristian** [reply]
> But if I load `localhost:8080/something`, should I get the `{ url: '/something' }`?

Yes, if you're opening that URL directly in your browser (as opposed to opening `/` and clicking a link to get to `/something`), and you prerender `/something`, then you should be seeing that. Are you not?

> Sorry, I'm a bit confused with how the prerender provider works

Basically, prerendering will output a data file (`/<route>/preact_prerender_data.js`) that the provider will then request when you open the page.

> Can I use the data from prerender-urls.js in dev?

Yes, but as mentioned, it will only run once. If you alter your prerender data, you'll need to restart the dev server.

**Reply by rschristian** [reply]
Maybe this demo will help: https://github.com/rschristian/preact-cli-prerender-data-provider-demo

**Reply by orangecoloured** [reply]
I don't think I'm getting proper props.
Here's a screencast. I'm refreshing the page being on `/graphics` but the url in the props is `/`.

https://github.com/preactjs/preact-cli/assets/3314891/8f1ac9b5-38cc-4093-96af-97f79ad37ff0

About the example you provided. I set up the `Provider` in the same way. But I don't see there the use of the hook `usePrerenderData`. From my understanding that's how we use the data from the provider in different routes. Is there a complete example?

**Reply by rschristian** [reply, marked as answer]
>  Here's a screencast. I'm refreshing the page being on `/graphics` but the url in the props is `/`.

That looks like you're not prerendering `/graphics` then.

Can you provide a reproduction? It's hard to diagnose an issue that I cannot see.

> From my understanding that's how we use the data from the provider in different routes.

That's one way to use it, sure. Just need a slight modification to a route to use it:

```js
// src/routes/home/index.js
import { usePrerenderData } from '@preact/prerender-data-provider';

const Home = (props) => {
    const [value] = usePrerenderData(props);
    ...
}
```

`value` will then have your prerender data for that route. `props` will come from the router, but just needs to be an object containing a URL.

> Is there a complete example?

[Best bet is in the CLI's test suite, here](https://github.com/preactjs/preact-cli/tree/master/packages/cli/tests/subjects/multiple-prerendering-with-provider) 

I should also be clear: `preact-cli` isn't maintained. I'm the only one who's done anything with it in years, and I really don't have time to invest anymore, especially seeing as how Webpack is effectively dead these days. If you've just picked up `preact-cli`, I'd definitely look elsewhere (we have a Vite initializer called [`create-preact`](https://npm.im/create-preact) which may be of interest).

This area especially, prerendering and how it bleeds into dev, was never something that was really fleshed out in a great way. It's pretty rough, as seen by the fact that you'll need to run a build before running `preact watch --prerender`.

**Reply by orangecoloured** [reply]
Thank you for a thorough and clear answer.

Yeah, I started using preact not long ago.
Alright, seems like I'll need to update my setup and, considering the preact-cli is unlikely to be maintained, I may switch to something else.

Thank you!

**Reply by rschristian** [reply]
I'm really sorry for the hassle.

We did update the docs to point at `create-preact` over `preact-cli` rather recently, not sure if you created your project before then or something. Right now it doesn't support prerendering out of the box but we're working to set that up. You can of course set it up yourself, it's quite simple, but will probably require looking into [`vite-plugin-ssr`](https://npm.im/vite-plugin-ssr).

Certainly let me know if i can be of any further help though.

**Reply by rschristian** [reply]
We should shortly have SSG support in `create-preact` (https://github.com/preactjs/create-preact/pull/1), giving an out-of-the-box similar experience to `preact-cli` FWIW.

**Reply by orangecoloured** [reply]
Is there a documentation I can refer to to migrate from `preact-cli` to `create-preact`?

**Reply by rschristian** [reply]
Not at the moment, no. Hopefully I'll be able to write some eventually.

Do you have any specific questions at the moment? I can try to answer them here.

**Reply by orangecoloured** [reply]
1. Is the folder structure same? Do I put my static files, like favicon, robots.txt to the same place?
2. I have a bunch of plugins pre/post-processing files running via webpack. Will it just work with the `create-preact`? So they are in the `webpack.config.js`, can I just leave them there?
3. How does prerendering work? Do I use the same `prerender-urls.js` with the same structure or should I do it some other way?

**Reply by rschristian** [reply]
> Is the folder structure same? Do I put my static files, like favicon, robots.txt to the same place?

No, best reference [Vite's docs on static asset handling](https://vitejs.dev/guide/assets.html#the-public-directory).

> I have a bunch of plugins pre/post-processing files running via webpack. Will it just work with the `create-preact`? So they are in the `webpack.config.js`, can I just leave them there?

No, none of them will be usable. As mentioned, `create-preact` is a Vite-based initializer. Vite uses the Rollup plugin interface (thanks to the Preact team, actually), not Webpack. You will need to find replacements for any that you use or create your own.

Here's [the docs on configuring Vite using plugins](https://vitejs.dev/guide/using-plugins.html).

> How does prerendering work? Do I use the same `prerender-urls.js` with the same structure or should I do it some other way?

No, `prerender-urls.js` cannot be used in any way. [`vite-plugin-ssr` will prerender all of the pages in the `pages` directory](https://vite-plugin-ssr.com/pre-rendering).

There is no real migration path for existing projects. Very few of the `preact-cli` ways of doing things will carry over.
